staticaddr/loopin/actions.go

```diff
@@ -0,0 +1,594 @@
+		f.ctx, nil, htlc.PkScript, defaultConfTarget,
+		int32(f.loopIn.InitiationHeight),
+	)
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	// Subscribe to new blocks.
+	blockChan, blockChanErr, err := f.cfg.ChainNotifier.RegisterBlockEpochNtfn(f.ctx) //nolint:lll
+	if err != nil {
+		return f.HandleError(err)
+	}
+
+	htlcConfirmed := false
+
+	for {
+		select {
+		case <-htlcConfChan:
+			htlcConfirmed = true
+
+		case err = <-htlcErrConfChan:
+			return f.HandleError(err)
+
+		case currentHeight := <-blockChan:
+			cancelInvoice := func() {
+				log.Errorf("timeout waiting for invoice to be "+
+					"paid, canceling invoice: %v",
+					f.loopIn.SwapHash)
+
+				err = f.cfg.InvoicesClient.CancelInvoice(
+					f.ctx, f.loopIn.SwapHash,
+				)
+				if err != nil {
+					log.Warnf("unable to cancel invoice "+
+						"for swap hash: %v %v",
+						f.loopIn.SwapHash, err)
+				}
+			}
+
+			htlcTimedOut := f.loopIn.isHtlcTimedOut(currentHeight)
+
+			if !htlcConfirmed {
+				// If the htlc hasn't confirmed but the timeout
+				// path opened up, and we didn't receive the
+				// swap payment we consider the swap attempt to
+				// be failed. We cancel the invoice and reset
+				// the deposits.
+				if htlcTimedOut {
+					cancelInvoice()
+
+					return OnPaymentTimedOut
+				}
+
+				// If the htlc hasn't timed out we have to wait
+				// for it to potentially do, and then sweep it.
+				continue
+			}
+
+			// If the htlc has confirmed and the timeout path has
+			// opened up we cancel the swap invoice and sweep the
+			// funds back to us.
+			if htlcTimedOut {
+				cancelInvoice()
+
+				err = f.cfg.DepositManager.TransitionDeposits(
+					f.loopIn.Deposits,
+					deposit.OnSweepingHtlcTimeout,
+					deposit.SweepHtlcTimeout,
+				)
+				if err != nil {
+					return f.HandleError(err)
+				}
+
+				return OnSweepHtlcTimeout
+			}
+
+		case err = <-blockChanErr:
+			return f.HandleError(err)
+
+		case err = <-invoiceErrChan:
+			return f.HandleError(err)
+
+		case update := <-invoiceUpdateChan:
+			switch update.State {
+			case invoices.ContractOpen:
+			case invoices.ContractAccepted:
+			case invoices.ContractSettled:
+				f.Debugf("received off-chain payment: %v "+
+					"update: %v", f.loopIn.SwapHash,
+					update.State)
+
+				return OnPaymentReceived
+
+			case invoices.ContractCanceled:
+				return f.HandleError(fmt.Errorf("invoice " +
+					"canceled"))
+
+			default:
+				return f.HandleError(fmt.Errorf("unexpected "+
+					"invoice state: %v", update.State))
+			}
+
+		case <-f.ctx.Done():
+			return f.HandleError(f.ctx.Err())
+		}
+	}
+}
+
+// SweepHtlcTimeoutAction is called if the server published the htlc tx without
+// paying the invoice. We wait for the timeout path to open up and sweep the
+// funds back to us.
+func (f *FSM) SweepHtlcTimeoutAction(_ fsm.EventContext) fsm.EventType {
+	for {
+		err := f.createAndPublishHtlcTimeoutSweepTx()
+		if err == nil {
+			break
```

Comment:
This only happens if there's a serious error, so I'd suggest adding detailed logging too (mentioning that it'll retry in one hour).